File: TeamCode/Proto/src/main/java/org/murraybridgebunyips/proto/Proto.java
package org.murraybridgebunyips.proto;

import org.murraybridgebunyips.bunyipslib.RobotConfig;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.DegreesPerSecond;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Inches;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.MetersPerSecond;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.MetersPerSecondPerSecond;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Millimeters;
import static org.murraybridgebunyips.bunyipslib.external.units.Units.Second;

import com.acmerobotics.roadrunner.control.PIDCoefficients;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.IMU;

import org.murraybridgebunyips.bunyipslib.Dbg;
import org.murraybridgebunyips.bunyipslib.Motor;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.DriveConstants;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.MecanumCoefficients;
import org.murraybridgebunyips.bunyipslib.roadrunner.drive.localizers.TwoWheelLocalizer;
import org.murraybridgebunyips.bunyipslib.roadrunner.util.Deadwheel;

/**
 * FTC 15215 INTO THE DEEP 2024-2025 robot configuration
 *
 * @author Lucas Bubner, 2024
 */
public class Proto extends RobotConfig {
    /**
     * Expansion 0: Front Left "fl"
     */
    public Motor frontLeft;
    /**
     * Expansion 1: Front Right "fr"
     */
    public Motor frontRight;
    /**
     * Expansion 2: Back Right "br"
     */
    public Motor backRight;
    /**
     * Expansion 3: Back Left "bl"
     */
    public Motor backLeft;
    /**
     * Control 3: Parallel Encoder "pe"
     */
    public Deadwheel parallelDeadwheel;
    /**
     * Control 2: Perpendicular Encoder "ppe"
     */
    public Deadwheel perpendicularDeadwheel;
    /**
     * Internally mounted on I2C C0 "imu"
     */
    public IMU imu;

    /**
     * RoadRunner drive constants
     */
    public DriveConstants driveConstants;
    /**
     * Dual deadwheel intrinsics
     */
    public TwoWheelLocalizer.Coefficients localizerCoefficients;
    /**
     * Mecanum coefficients
     */
    public MecanumCoefficients mecanumCoefficients;

    @Override
    protected void onRuntime() {
        // Base is from GLaDOS
        imu = getHardware("imu", IMU.class, (d) -> {
            boolean init = d.initialize(new IMU.Parameters(
                    new RevHubOrientationOnRobot(
                            RevHubOrientationOnRobot.LogoFacingDirection.UP,
                            RevHubOrientationOnRobot.UsbFacingDirection.LEFT
                    )
            ));
            if (!init) Dbg.error("imu failed init");
        });
        frontLeft = getHardware("fl", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.FORWARD));
        frontRight = getHardware("fr", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));
        backRight = getHardware("br", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));
        backLeft = getHardware("bl", Motor.class, (d) -> d.setDirection(DcMotorSimple.Direction.REVERSE));

        // REV Through Bore Encoders
        parallelDeadwheel = getHardware("pe", Deadwheel.class,
                (d) -> d.setDirection(Deadwheel.Direction.FORWARD));
        perpendicularDeadwheel = getHardware("ppe", Deadwheel.class,
                (d) -> d.setDirection(Deadwheel.Direction.FORWARD));
        localizerCoefficients = new TwoWheelLocalizer.Coefficients.Builder()
                .setTicksPerRev(8192)
                .setGearRatio(1)
                .setWheelRadius(Millimeters.of(60).divide(2))
                // TODO: calibrate
//                .setXMultiplier(100.0 / 134.5)
//                .setYMultiplier(100.0 / 134.5)
                .setParallelX(Inches.zero())
                .setParallelY(Inches.of(-2))
                .setPerpendicularX(Inches.one())
                .setPerpendicularY(Inches.of(2.5))
                .setOverflowCompensation(true)
                .build();

        // RoadRunner configuration
        // TODO: not tuned for robot construction due to weight
        driveConstants = new DriveConstants.Builder()
                .setTicksPerRev(28)
                .setMaxRPM(458)
                .setRunUsingEncoder(false)
                .setWheelRadius(Millimeters.of(75).divide(2))
                .setGearRatio(1.0 / 13.1)
                .setTrackWidth(Inches.of(20.5))
                .setMaxVel(MetersPerSecond.of(1.04))
                .setMaxAccel(MetersPerSecondPerSecond.of(1.04))
                .setMaxAngVel(DegreesPerSecond.of(175))
                .setMaxAngAccel(DegreesPerSecond.per(Second).of(175))
                .setKV(0.01395)
                .setKStatic(0.06311)
                .setKA(0.0015)
                .build();
        mecanumCoefficients = new MecanumCoefficients.Builder()
                // TODO: calibrate
                .setLateralMultiplier(60.0 / 59.846666)
                .setTranslationalPID(new PIDCoefficients(8, 0, 0))
                .setHeadingPID(new PIDCoefficients(10, 0, 0))
                .build();
    }
}

File: TeamCode/Proto/src/main/java/au/edu/sa/mbhs/studentrobotics/ftc15215/proto/Proto.kt
package au.edu.sa.mbhs.studentrobotics.ftc15215.proto

import au.edu.sa.mbhs.studentrobotics.bunyipslib.BunyipsOpMode
import au.edu.sa.mbhs.studentrobotics.bunyipslib.RobotConfig
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.TrapezoidProfile
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.ff.ElevatorFeedforward
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.control.pid.PController
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Unit.Companion.of
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.Inches
import au.edu.sa.mbhs.studentrobotics.bunyipslib.external.units.Units.InchesPerSecond
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.Motor
import au.edu.sa.mbhs.studentrobotics.bunyipslib.hardware.ProfiledServo
import au.edu.sa.mbhs.studentrobotics.bunyipslib.localization.TwoWheelLocalizer
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.DriveModel
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.MecanumGains
import au.edu.sa.mbhs.studentrobotics.bunyipslib.roadrunner.parameters.MotionProfile
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.DualServos
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.HoldableActuator
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.Switch
import au.edu.sa.mbhs.studentrobotics.bunyipslib.subsystems.drive.MecanumDrive
import au.edu.sa.mbhs.studentrobotics.ftc15215.proto.components.LinkedLift
import com.acmerobotics.roadrunner.ftc.LazyImu
import com.acmerobotics.roadrunner.ftc.RawEncoder
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.Servo

/**
 * FTC 15215 INTO THE DEEP 2024-2025 robot configuration
 *
 * @author Lucas Bubner, 2024
 */
class Proto : RobotConfig() {
    /**
     * Non-subsystem hardware instances.
     */
    val hw = Hardware()

    /**
     * RoadRunner Mecanum Drive with Two-Wheel Localization.
     */
    lateinit var drive: MecanumDrive

    /**
     * Dual servos for the claw mechanism.
     */
    lateinit var claws: DualServos

    /**
     * `claws` rotation mechanism.
     */
    lateinit var clawRotator: Switch

    /**
     * Vertical lift for the claw mechanism (`claws` and `clawRotator`).
     */
    lateinit var clawLift: HoldableActuator

    /**
     * Linked ascension mechanism for the claw lift.
     */
    lateinit var ascent: LinkedLift

    override fun onRuntime() {
        // Base is from GLaDOS
        hw.imu = getLazyImu(
            orientationOnRobot = RevHubOrientationOnRobot(
                RevHubOrientationOnRobot.LogoFacingDirection.UP,
                RevHubOrientationOnRobot.UsbFacingDirection.LEFT
            )
        )
        hw.fl = getHardware("fl", DcMotorEx::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
        }
        hw.bl = getHardware("bl", DcMotorEx::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
        }
        hw.br = getHardware("br", DcMotorEx::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
        }
        hw.fr = getHardware("fr", DcMotorEx::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
        }

        // REV Through Bore Encoders attached on ports 0 and 3 with the motors
        hw.pe = getHardware("fr", RawEncoder::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
        }
        hw.ppe = getHardware("br", RawEncoder::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
        }

        // End effectors
        hw.leftClaw = getHardware("lc", Servo::class.java) {
            it.direction = Servo.Direction.REVERSE
            it.scaleRange(0.6, 1.0)
        }
        hw.rightClaw = getHardware("rc", Servo::class.java) {
            it.direction = Servo.Direction.FORWARD
            it.scaleRange(0.0, 0.4)
        }
        hw.clawRotator = getHardware("cr", ProfiledServo::class.java) {
            it.setConstraints(TrapezoidProfile.Constraints(Constants.cr_v, Constants.cr_a))
            it.direction = Servo.Direction.REVERSE
            it.scaleRange(0.6, 1.0)
        }

        hw.clawLift = getHardware("cl", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.REVERSE
            val p = PController(Constants.cl_kP)
            val ff = ElevatorFeedforward(0.0, Constants.cl_kG, 0.0, 0.0, { 0.0 }, { 0.0 })
            val c = p.compose(ff) { a, b -> a + b }
            it.runToPositionController = c
            BunyipsOpMode.ifRunning { o -> o.onActiveLoop({ c.setCoefficients(Constants.cl_kP, 0.0, 0.0, 0.0, 0.0, Constants.cl_kG, 0.0, 0.0) }) }
        }

        hw.leftAscent = getHardware("la", Motor::class.java) {
            it.direction = DcMotorSimple.Direction.FORWARD
            val c = PController(Constants.a_kP)
            it.runToPositionController = c
            BunyipsOpMode.ifRunning { o -> o.onActiveLoop({ c.setCoefficients(Constants.a_kP, 0.0, 0.0, 0.0) })}
        }
        hw.rightAscent = getHardware("ra", Motor::class.java) {
            // Child of leftAscent, controller configured in LinkedLift
            it.direction = DcMotorSimple.Direction.REVERSE
        }

        // RoadRunner drivebase configuration
        val dm = DriveModel.Builder()
            .setDistPerTick(123.5 of Inches, 134005.0)
            .setLateralInPerTick(0.0007201181372108113)
            .setTrackWidthTicks(15480.209096658593)
            .build()
        val mp = MotionProfile.Builder()
            .setMaxWheelVel(40.0 of InchesPerSecond)
            .setKs(0.93)
            .setKv(0.00022)
            .setKa(0.00001)
            .build()
        val mg = MecanumGains.Builder()
            .setAxialGain(2.5)
            .setLateralGain(2.5)
            .setHeadingGain(4.0)
            .build()
        val twl = TwoWheelLocalizer.Params.Builder()
            .setParYTicks(-2109.055102501924)
            .setPerpXTicks(-5279.813561122253)
            .build()

        drive = MecanumDrive(dm, mp, mg, hw.fl, hw.bl, hw.br, hw.fr, hw.imu, hardwareMap.voltageSensor)
            .withLocalizer(TwoWheelLocalizer(dm, twl, hw.pe, hw.ppe, hw.imu?.get()))
            .withName("Drive")
        claws = DualServos(hw.leftClaw, hw.rightClaw)
            .withName("Claws")
        clawRotator = Switch(hw.clawRotator)
            .withName("Claw Rotator")
        clawLift = HoldableActuator(hw.clawLift)
            .enableUserSetpointControl { dt -> dt * Constants.cl_TPS }
            .withName("Claw Lift")
        ascent = LinkedLift(hw.leftAscent, hw.rightAscent)
            .withName("Ascender")
    }

    /**
     * Raw hardware instances.
     */
    class Hardware {
        /**
         * Expansion 0: Front Left "fl"
         */
        var fl: DcMotorEx? = null

        /**
         * Expansion 1: Front Right "fr"
         */
        var fr: DcMotorEx? = null

        /**
         * Expansion 2: Back Right "br"
         */
        var br: DcMotorEx? = null

        /**
         * Expansion 3: Back Left "bl"
         */
        var bl: DcMotorEx? = null

        /**
         * Expansion 3: Parallel Encoder "fr" (on motor)
         */
        var pe: RawEncoder? = null

        /**
         * Expansion 0: Perpendicular Encoder "br" (on motor)
         */
        var ppe: RawEncoder? = null

        /**
         * Internally mounted on I2C C0 "imu"
         */
        var imu: LazyImu? = null

        /**
         * Control S1: Left Claw "lc"
         */
        var leftClaw: Servo? = null

        /**
         * Control S0: Right Claw "rc"
         */
        var rightClaw: Servo? = null

        /**
         * Control S2: Claw Rotator "cr"
         */
        var clawRotator: Servo? = null

        /**
         * Control 1: Claw Lift "cl"
         */
        var clawLift: Motor? = null

        /**
         * Control 3: Left Ascent "la"
         */
        var leftAscent: Motor? = null

        /**
         * Control 0: Right Ascent "ra"
         */
        var rightAscent: Motor? = null
    }
}